reconcile/aws_version_sync/integration.py

```diff
@@ -0,0 +1,297 @@
+        timeout: int,
+        prom_get_func: Callable = prom_get,
+    ) -> list[ExternalResource]:
+        metrics: list[ExternalResource] = []
+        # compile a list of all RDS metrics from all AWS resource exporter clusters; ignore duplicated clusters
+        for cluster in uniquify(key=lambda c: c.name, items=clusters):
+            token = (
+                self.secret_reader.read_secret(cluster.automation_token)
+                if cluster.automation_token
+                else None
+            )
+            try:
+                metrics += [
+                    ExternalResource(
+                        provider="aws",
+                        provisioner=ExternalResourceProvisioner(
+                            uid=m["aws_account_id"]
+                        ),
+                        resource_provider="rds",
+                        resource_identifier=m["dbinstance_identifier"],
+                        resource_engine=m["engine"],
+                        resource_engine_version=m["engine_version"],
+                    )
+                    for m in prom_get_func(
+                        cluster.prometheus_url,
+                        {"query": "aws_resources_exporter_rds_engineversion"},
+                        token=token,
+                        timeout=timeout,
+                    )
+                ]
+            except Exception as e:
+                logging.error(f"Failed to get metrics for cluster {cluster.name}: {e}")
+        return metrics
+
+    def get_external_resource_specs(
+        self,
+        gql_get_resource_func: Callable,
+        namespaces: Iterable[NamespaceV1],
+        supported_providers: Iterable[str],
+    ) -> list[ExternalResource]:
+        external_resources: list[ExternalResource] = []
+        _defaults_cache: dict[str, Any] = {}
+        for ns in namespaces:
+            for external_resource in ns.external_resources or []:
+                if not isinstance(
+                    external_resource, NamespaceTerraformProviderResourceAWSV1
+                ):
+                    continue
+                # make mypy happy
+                assert isinstance(external_resource.provisioner, AWSAccountV1)
+
+                for resource in external_resource.resources or []:
+                    if resource.provider.lower() not in supported_providers:
+                        continue
+
+                    # make mypy happy
+                    assert isinstance(resource, NamespaceTerraformResourceRDSV1)
+
+                    # get/set the defaults file values from/to cache
+                    values = (
+                        _defaults_cache.setdefault(
+                            resource.defaults,
+                            # retrieve the external resource spec values
+                            get_values(gql_get_resource_func, resource.defaults),
+                        )
+                        if resource.defaults
+                        else {}
```

Comment:
🤦 you're right. fixed in [50c07fa](https://github.com/app-sre/qontract-reconcile/pull/3934/commits/50c07fa313fdafd5c4d99699c93209d0ea4a5668)

Reply:
since `get_values` always return a dict, can use `:=` and none check here:

```python
                    if resource.defaults:
                        if (values := _defaults_cache.get(resource.defaults)) is None:
                            values = get_values(
                                gql_get_resource_func, resource.defaults
                            )
                            _defaults_cache[resource.defaults] = values
```

Reply:
Applied in [c0ed7a7](https://github.com/app-sre/qontract-reconcile/pull/3934/commits/c0ed7a765712beb5f062b6d59abaa7b86fa2e8ed)